api/commands/run.ts

```diff
@@ -112,24 +112,54 @@ export class Runner {
   }
 
   private async executeAllActionsReadyForExecution() {
-    const allSuccessfulActions = this.result.actions.filter(isSuccessfulAction).map(fn => fn.name);
-    const isReadyForExecution = (action: dataform.IExecutionAction) => {
-      for (const dependency of action.dependencies) {
-        if (!allSuccessfulActions.includes(dependency)) {
-          return false;
-        }
-      }
-      return true;
-    };
-    const readyForExecutionActions = this.pendingActions.filter(isReadyForExecution);
-    this.pendingActions = this.pendingActions.filter(action => !isReadyForExecution(action));
-    return Promise.all(
-      readyForExecutionActions.map(async action => {
-        this.result.actions.push(await this.executeAction(action));
-        await this.triggerChange();
-        await this.executeAllActionsReadyForExecution();
-      })
+    // If the run has been cancelled, skip all pending actions.
+    if (this.cancelled) {
+      const allPendingActions = this.pendingActions;
+      this.pendingActions = [];
+      await Promise.all(
+        allPendingActions.map(async pendingAction => this.skipAction(pendingAction))
+      );
+      return;
+    }
+
+    // Determine what actions we can execute.
+    const allDependenciesHaveExecutedSuccessfully = allDependenciesHaveBeenExecuted(
+      this.result.actions.filter(isSuccessfulAction)
+    );
+    const executableActions = this.pendingActions.filter(allDependenciesHaveExecutedSuccessfully);
+    this.pendingActions = this.pendingActions.filter(
+      action => !allDependenciesHaveExecutedSuccessfully(action)
     );
+
+    // Determine what actions we should skip (necessarily excluding executable actions).
+    const allDependenciesHaveExecuted = allDependenciesHaveBeenExecuted(this.result.actions);
+    const skippableActions = this.pendingActions.filter(allDependenciesHaveExecuted);
+    this.pendingActions = this.pendingActions.filter(
+      action => !allDependenciesHaveExecuted(action)
+    );
+
+    await Promise.all([
+      Promise.all(skippableActions.map(async skippableAction => this.skipAction(skippableAction))),
+      Promise.all(
+        executableActions.map(async executableAction =>
+          this.onActionExecutionComplete(await this.executeAction(executableAction))
+        )
+      )
+    ]);
+  }
+
+  private async onActionExecutionComplete(executedAction: dataform.IExecutedAction) {
+    this.result.actions.push(executedAction);
+    await this.triggerChange();
+    await this.executeAllActionsReadyForExecution();
```

Comment:
Hmmm. AIUI, you build up this massive dependency tree of promises that's the inverse of the dependecy graph, that's going to recursively come all the way back up to the top when it's done?
Feels like when this goes wrong it's going to be very hard to debug / reason about.
Personally I thought the for loop was more readable. (Not going to block on this though).

Reply:
unless I misunderstood the previous code, this code is very similar to the loop, which was itself recursive (or at least, triggered calls to itself).

i could be misunderstanding your comment, but I don't think it works like that. this code does not build an inverse of the dependency graph, it's very like the old code in that upon each iteration it (a) finds and executes all executable actions, (b) skips all skippable actions, (c) triggers another iteration of all of the above, until the list of pending actions is exhausted.